Kitodo/src/main/java/org/kitodo/services/image/ImageGenerator.java

```diff
@@ -0,0 +1,423 @@
+     * @param outputs
+     *            output folders to generate to
+     */
+    public ImageGenerator(Subfolder sourceFolder, GenerationMode mode, Collection<Subfolder> outputs) {
+        this.sourceFolder = sourceFolder;
+        this.mode = mode;
+        this.outputs = outputs;
+        this.state = ImageGeneratorStep.LIST_SOURCE_FOLDER;
+        this.sources = Collections.emptyList();
+        this.toBeGenerated = new LinkedList<>();
+    }
+
+    /**
+     * Appends the element to the list of elements to be generated.
+     *
+     * @param elementToAppend
+     *            element to be appended to the list
+     */
+    void addToToBeGenerated(Pair<Pair<String, URI>, List<Subfolder>> elementToAppend) {
+        toBeGenerated.add(elementToAppend);
+    }
+
+    /**
+     * Invokes the image management interface method that creates a derivative.
+     * What kind of derivative should be created is determined from the folder
+     * configuration.
+     *
+     * @param sourceImage
+     *            address of the source image from which the derivative is to be
+     *            calculated.
+     * @param imageProperties
+     *            configuration for the target image
+     * @param imageFileFormat
+     *            image file format to create
+     * @param destinationImage
+     *            image file to write
+     * @throws IOException
+     *             if an underlying disk operation fails
+     */
+    private static void createDerivative(URI sourceImage, Folder imageProperties, ImageFileFormat imageFileFormat,
+            URI destinationImage) throws IOException {
+
+        ImageManagementInterface imageGenerator 
+            = new KitodoServiceLoader<ImageManagementInterface>(ImageManagementInterface.class).loadModule();
+        imageGenerator.createDerivative(sourceImage, imageProperties.getDerivative().get(), destinationImage,
+            imageFileFormat);
+    }
+
+    /**
+     * Generates a set of derivatives.
+     *
+     * @param instruction
+     *            Instruction, which pictures are to be generated. Left: image
+     *            source, right: destination folder. The image source consists
+     *            of the canonical part of the file name and the resolved file
+     *            name for the source image. The canonical part of the file name
+     *            is needed to calculate the corresponding file name in the
+     *            destination folder. The type of derivative to be generated is
+     *            defined in the properties of the destination folder.
+     */
+    void createDerivatives(Pair<Pair<String, URI>, List<Subfolder>> instruction) {
+        try {
+            Pair<String, URI> imageSource = instruction.getLeft();
+            for (Subfolder destinationFolder : instruction.getRight()) {
+                generateDerivative(imageSource.getValue(), destinationFolder, imageSource.getKey());
+            }
+        } catch (IOException e) {
```

Comment:
Why do you replace the IOException with an UndeclaredThrowableException?

Reply:
Because the method is called within a consumer, and the interface consumer does not allow a checked exception. I could also throw the exception here and catch it at the next higher level, but I would have to do it in one of the two places, and that method is shorter, so I did it here so that both methods stay clearer.